cloudinary-core/src/main/java/com/cloudinary/Api.java

```diff
@@ -552,4 +560,81 @@ private ApiResponse updateResourcesAccessMode(String accessMode, String byKey, O
         return callApi(HttpMethod.POST, uri, params, options);
     }
 
+    /**
+     * Add a new metadata field definition
+     * @param field The field to add.
+     * @return A map representing the newlay added field.
+     * @throws Exception
+     */
+    public ApiResponse addMetadataField(AbstractMetadataField field) throws Exception {
+        return callApi(HttpMethod.POST, Collections.singletonList("metadata_fields"),
+                ObjectUtils.toMap(field), ObjectUtils.asMap ("content_type", "json"));
+    }
+
+    /**
+     * List all the metadata field definitions (structure, not values)
+     * @return A map containing the list of field definitions.
+     * @throws Exception
+     */
+    public ApiResponse listMetadataFields() throws Exception {
+        return callApi(HttpMethod.GET, Collections.singletonList("metadata_fields"), Collections.<String, Object>emptyMap(), Collections.emptyMap());
+    }
+
+    /**
+     * Get a metadata field definition by id
+     * @param externalId The if of the field to retrive
+     * @return The fields definitions.
+     * @throws Exception
+     */
+    public ApiResponse metadataFieldByFieldId(String externalId) throws Exception {
+        return callApi(HttpMethod.GET, Arrays.asList("metadata_fields", externalId), Collections.<String, Object>emptyMap(), Collections.emptyMap());
+    }
+
+    /**
+     * Update the definitions of a single metadata field.
+     * @param externalId The id of the field to update
+     * @param field The field definition
+     * @return The updated fields definition.
+     * @throws Exception
+     */
+    public ApiResponse updateMetadataField(String externalId, AbstractMetadataField field) throws Exception {
+        List<String> uri = Arrays.asList("metadata_fields", externalId);
+        return callApi(HttpMethod.PUT, uri, ObjectUtils.toMap(field), Collections.singletonMap("content_type", "json"));
+    }
+
+    /**
+     * Update the datasource entries for a given field
+     * @param fieldExternalId The id of the field to update
```

Comment:
Inconsistency - sometimes it is called externalId and sometimes fieldExternalId. 
I suggest using fieldId (not sure what the external represents)

Reply:
When you send an ID to a function about a field you an omit the 'field' for brevity. When you need to provide an id for a field in the context of another API (e.g. datasource, that has it's own id() you must include the 'field' in the parameter name for clarity.

I can use 'fieldExternalId' everywhere for consistency. The word external is part of the spec, that's the key used in all the requests and responses, we have to use it.

Reply:
Regarding the fieldExternalId vs. externalId, got it. 

Regarding the external, 
I understand that this is what the server is using, but I don't understand why they are using it. 
In any case, why do we have to be aligned with the server naming?

Reply:
I had a long discussion about it with docs, for them it's real bad if the API docs use different names than the SDK docs so we're using the same names. We can't open it up tomorrow to if you want, for us it's a quick easy fix...

Reply:
Cool. let's talk about it